src/Hamiltonians/Molecular.jl

```diff
@@ -0,0 +1,773 @@
+
+function Base.iterate(
+    iter::MolecularHamiltonianOffDiagonals,
+    state::MolecularHamiltonianOffDiagonalsIteratorState
+)
+    na, nb = state.excitations_per_channel
+    if (na, nb) == (0, 1) # single electron in beta channel
+        if is_void_state(state) || is_invalid_state(iter, state)
+            return iterate(
+                iter, MolecularHamiltonianOffDiagonalsIteratorState((1, 0), (1, 0), (1, 0))
+            )
+        end
+        r = one_electron_excitation(2, iter.address, iter.op, iter.modes, state)
+        nstate = one_electron_excitation_next(iter, state)
+        return r, nstate
+    elseif (na, nb) == (1, 0) # single electron excitation in alpha channel
+        if is_void_state(state) || is_invalid_state(iter, state)
+            return iterate(
+                iter, MolecularHamiltonianOffDiagonalsIteratorState((0, 2), (1, 2), (1, 2))
+            )
+        end
+        r = one_electron_excitation(1, iter.address, iter.op, iter.modes, state)
+        nstate = one_electron_excitation_next(iter, state)
+        return r, nstate
+    elseif (na, nb) == (0, 2) # 2 electron excitation in beta channel
+        if is_void_state(state) || is_invalid_state(iter, state)
+            return iterate(
+                iter, MolecularHamiltonianOffDiagonalsIteratorState((2, 0), (1, 2), (1, 2))
+            )
+        end
+        r = two_electron_excitation(2, iter.address, iter.op, iter.modes, state)
+        nstate = two_electron_excitation_next(iter, state)
+        return r, nstate
+    elseif (na, nb) == (2, 0) # 2 electron excitation in alpha channel
+        if is_void_state(state) || is_invalid_state(iter, state)
+            return iterate(
+                iter, MolecularHamiltonianOffDiagonalsIteratorState((1, 1), (1, 1), (1, 1))
+            )
+        end
+        r = two_electron_excitation(1, iter.address, iter.op, iter.modes, state)
+        nstate = two_electron_excitation_next(iter, state)
+        return r, nstate
+    elseif (na, nb) == (1, 1)
+        if is_void_state(state) || is_invalid_state(iter, state)
+            return nothing
+        end
+        r = one_one_electron_excitation(iter.address, iter.op, iter.modes, state)
+        nstate = one_one_electron_excitation_next(iter, state)
+        return r, nstate
+    end
+end
+
+function Base.length(iter::MolecularHamiltonianOffDiagonals)
+    return length(iter.modes.occupied[1]) * length(iter.modes.unoccupied[1]) + # 1-alpha
+           length(iter.modes.occupied[2]) * length(iter.modes.unoccupied[2]) + # 1-beta
+           binomial(length(iter.modes.occupied[1]), 2) * 
+           binomial(length(iter.modes.unoccupied[1]), 2) + # 2-alpha
+           binomial(length(iter.modes.occupied[2]), 2) * 
+           binomial(length(iter.modes.unoccupied[2]), 2) + # 2-beta
+           (length(iter.modes.occupied[1]) * length(iter.modes.unoccupied[1]) *
+            length(iter.modes.occupied[2]) * length(iter.modes.unoccupied[2])) # 1-alpha, 1-beta
```

Comment:
Isn't this information already contained in the `ind_cap` field and just needs to be added?